### PR TITLE
Flatten Scheduler Struct

### LINKED ISSUE #71860: [scheduler cleanup phase 2]: Flatten Top Level Scheduler Struct
**PLEASE COORDINATE WITH SIG-SCHEDULING BEFORE WORKING ON THIS ISSUE.** We want to avoid the situation where two people accidentally start work on the same fix.

Main tracking issue: https://github.com/kubernetes/kubernetes/issues/68951

If you contributed in phase 1, please give others a chance before claiming this issue. You can also help out by reviewing the phase 2 PRs.

Currently, the type [`scheduler.Scheduler`](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/scheduler.go) is a struct with one field. (`factory.Config`) It is misleading for all of the scheduler’s config *and state* to be stored in a type called "config".

Instead, we can merge these two types, keeping only `scheduler.Scheduler`. Any function returning a Config will return a Scheduler instead. Integration tests will need to be fixed to accommodate this change.

/sig scheduling
/kind cleanup
/help
/good-first-issue

### PR DESCRIPTION
**Pull request by ahmad-diaa**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR aims to flatten `scheduler.Scheduler` type and make the scheduler independent of the the `configFactory` type.
**Which issue(s) this PR fixes**:

Fixes #71860, #72547

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by vllry** [comment]
@ahmad-diaa would you mind addressing the merge conflicts?

/assign @wojtek-t

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -139,36 +197,77 @@ func New(client clientset.Interface,
...
+		storageClassLister = storageClassInformer.Lister()
+	}
+
+	podQueue := internalqueue.NewSchedulingQueue(stopEverything)
+	// Setup volume binder
+	volumeBinder := volumebinder.NewVolumeBinder(client, nodeInformer, pvcInformer, pvInformer, storageClassInformer, time.Duration(options.bindTimeoutSeconds)*time.Second)
+	scheduledPodsHasSynced := podInformer.Informer().HasSynced
+	// ScheduledPodLister is something we provide to plug-in functions that
+	// they may need to call.
+	// scheduledPodLister := assignedPodLister{podInformer.Lister()}
```
Comment (misterikkit) [inline pkg/scheduler/scheduler.go]: Can this be deleted? I believe this is referring to the old concept of factory plugins and not the new framework plugins.

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -55,12 +69,56 @@ const (
 // Scheduler watches for new unscheduled pods. It attempts to find
 // nodes that they fit on and writes bindings back to the api server.
 type Scheduler struct {
```
Comment (misterikkit) [inline pkg/scheduler/scheduler.go]: I'd like to find out which of these fields can be made lower case without breaking anything. I'm OK with making that change in a follow up PR.

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -139,36 +197,77 @@ func New(client clientset.Interface,
...
+	)
+	debugger.ListenForSignal(stopEverything)
+
+	go func() {
+		<-stopEverything
+		podQueue.Close()
+	}()
+
+	// initiate pluginFactoryArgs
+	pluginArgs := &factory.PluginFactoryArgs{
```
Comment (misterikkit) [inline pkg/scheduler/scheduler.go]: This isn't used until line 353. Can we move the initialization closer to that line?

File: test/integration/daemonset/daemonset_test.go [hunk 4]
```diff
@@ -97,47 +98,34 @@ func setupScheduler(
...
-	})
-	schedulerConfig, err := schedulerConfigFactory.Create()
-	if err != nil {
-		t.Fatalf("Couldn't create scheduler config: %v", err)
-	}
-
+	eventBroadcaster := record.NewBroadcaster()
+	defaultBindTimeout := int64(30)
+	defaultSource := "DefaultProvider"
 	// TODO: Replace NewFromConfig and AddAllEventHandlers with scheduler.New() in
```
Comment (misterikkit) [inline test/integration/daemonset/daemonset_test.go]: Please delete this TODO comment.

File: pkg/scheduler/scheduler.go [hunk 5]
```diff
@@ -242,31 +559,40 @@ func initPolicyFromConfigMap(client clientset.Interface, policyRef *kubeschedule
 func NewFromConfig(config *factory.Config) *Scheduler {
```
Comment (misterikkit) [inline pkg/scheduler/scheduler.go]: Are there any remaining callers to this function? Can we delete it yet?

**Review by misterikkit — COMMENTED** [review]
Overall, I love this PR.

There are lots of cleanup tasks that will be more obvious and easy to tackle after this merges.

Also, I think we should be able to delete `factory.Config` and the whole "configurator" thing. If you'd like that to be a separate PR, I'm OK with that.

File: pkg/scheduler/scheduler.go (hunk 5 above)
Comment (ahmad-diaa) [reply pkg/scheduler/scheduler.go]: There are still some callers in the tests. I will delete it after replacing all of them.

**Comment by ahmad-diaa** [comment]
@misterikkit I would be happy to open follow up PRs to address the following as this PR already has lots of changes.

> I'd like to find out which of these fields can be made lower case without breaking anything. I'm OK with making that change in a follow up PR.

> Also, I think we should be able to delete `factory.Config` and the whole "configurator" thing. If you'd like that to be a separate PR, I'm OK with that.

**Comment by ahmad-diaa** [comment]
@misterikkit PTAL, thanks!

**Review by draveness — COMMENTED** [review]
/close

in favor of #81207

**Pull request closed without merging**